django/db/models/base.py

```diff
@@ -1926,11 +1936,14 @@ def _check_constraints(cls, databases):
                         id='models.W038',
                     )
                 )
-            fields = (
-                field
-                for constraint in cls._meta.constraints if isinstance(constraint, UniqueConstraint)
-                for field in constraint.fields
-            )
+            fields = []
+            for constraint in cls._meta.constraints:
+                if isinstance(constraint, UniqueConstraint):
+                    fields += constraint.fields
+                    if isinstance(constraint.condition, Q):
+                        fields += cls._get_check_or_condition_fields(constraint.condition.children)
+                elif isinstance(constraint, CheckConstraint) and isinstance(constraint.check, Q):
+                    fields += cls._get_check_or_condition_fields(constraint.check.children)
```

Comment:
Might want to make `fields` a set to avoid having `_check_local_fields` run multiple time for the same field.